Add Foundry and implement SolidDaoManagement and SCTERC20 unit tests

freitasgouvea opened the pull request:
- Configure and add Foundry tool
- Implement SolidDaoManagementTest unit tests (20 tests)
- Implement SCTERC20Test unit tests (32 tests)
- Update package and readme

Closed #21 
Closed #22

jeffprestes commented on test/SolidDaoManagement.t.sol:
```diff
@@ -0,0 +1,222 @@
+    function testPullGovernor() public {
+        vm.prank(governorAddress);
+        vm.expectEmit(true, true, true, true);
+        emit GovernorPushed(governorAddress, newGovernorAddress, false);
```
Would not be better to emit this event this after the action method tested (in this case pushGovernor) be performed ?

jeffprestes commented:
Also, when I had tried to follow the `Run unit tests` instructions in the README.md in a new machine I got an error when I executed `npm run test` command:

```bash
 npm run test

> solid-world-dao-contracts@1.0.0 test
> forge test -vvv

[⠊] Compiling...
[⠒] Unable to resolve import: "forge-std/Test.sol" with remappings:
    contracts/=/Users/jeffprestes/projetos/go/src/github.com/solid-world/solid-world-dao-contracts/contracts/
[⠢] Unable to resolve import: "forge-std/Test.sol" with remappings:
    contracts/=/Users/jeffprestes/projetos/go/src/github.com/solid-world/solid-world-dao-contracts/contracts/
[⠰] Compiling 32 files with 0.8.13
[⠔] Solc finished in 26.87ms
Error:
   0: Compiler run failed
      ParserError: Source "forge-std/Test.sol" not found: File not found. Searched the following locations: "".
       --> /Users/jeffprestes/projetos/go/src/github.com/solid-world/solid-world-dao-contracts/test/SCTERC20.t.sol:4:1:
        |
      4 | import "forge-std/Test.sol";
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^



      ParserError: Source "forge-std/Test.sol" not found: File not found. Searched the following locations: "".
       --> /Users/jeffprestes/projetos/go/src/github.com/solid-world/solid-world-dao-contracts/test/SolidDaoManagement.t.sol:4:1:
        |
      4 | import "forge-std/Test.sol";
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^


   0:

Location:
   cli/src/cmd/utils.rs:43
 ``` 
 
Please, add more details in this documentation section, including git modules updates and also Solidity contract compilations steps that I guess it will be missed.

freitasgouvea commented:
> ```shell
> Unable to resolve import: "forge-std/Test.sol" with remappings:
> ```

Yes, I missed one step, before run tests we need to register lib/forge-std. I updated the readme with this step. Thanks

jeffprestes commented:
Now the tests are running but 4 of them are breaking:

```bash
╰─ npm run test

> solid-world-dao-contracts@1.0.0 test
> forge test -vvv

[⠊] Compiling...
[⠒] Compiling 7 files with 0.8.13
[⠢] Solc finished in 2.00s
Compiler run successful (with warnings)

Running 20 tests for test/SolidDaoManagement.t.sol:SolidDaoManagementTest
[PASS] testGovernorAddress() (gas: 9909)
[PASS] testGuardianAddress() (gas: 9841)
[PASS] testPolicyAddress() (gas: 9799)
[PASS] testPullGovernor() (gas: 56316)
[PASS] testPullGuardian() (gas: 60349)
[PASS] testPullPolicy() (gas: 60296)
[PASS] testPullVault() (gas: 60468)
[PASS] testPushGovernor() (gas: 49631)
[PASS] testPushGuardian() (gas: 51667)
[PASS] testPushPolicy() (gas: 51675)
[PASS] testPushVault() (gas: 51696)
[PASS] testUnauthorizedPullGovernor() (gas: 47838)
[PASS] testUnauthorizedPullGuardian() (gas: 51762)
[PASS] testUnauthorizedPullPolicy() (gas: 51795)
[PASS] testUnauthorizedPullVault() (gas: 51818)
[PASS] testUnautorizedPushGovernor() (gas: 26150)
[PASS] testUnautorizedPushGuardian() (gas: 26194)
[PASS] testUnautorizedPushPolicy() (gas: 26193)
[PASS] testUnautorizedPushVault() (gas: 26235)
[PASS] testVaultAddress() (gas: 9951)
Test result: ok. 20 passed; 0 failed; finished in 10.45ms

Running 32 tests for test/SCTERC20.t.sol:SCTERC20Test
[PASS] testApproveFromZeroAddress() (gas: 15356)
[PASS] testApproveToZeroAddress() (gas: 15390)
[PASS] testApproveWithSuccess() (gas: 44516)
[PASS] testBalanceOf() (gas: 25531)
[PASS] testBurnFromInsufficientAllowance() (gas: 51874)
[PASS] testBurnFromInsufficientBalance() (gas: 40931)
[PASS] testBurnFromWithSuccess() (gas: 46932)
[PASS] testBurnWhenAmmountExceedsBalance() (gas: 16964)
[PASS] testBurnWithSuccess() (gas: 22458)
[PASS] testBurnWithZeroBalance() (gas: 16965)
[PASS] testBurnZeroAddress() (gas: 12538)
[PASS] testDecimals() (gas: 5502)
[PASS] testDecreaseAllowancewBellowZero() (gas: 45989)
[PASS] testDecreaseAllowancewWithSuccess() (gas: 49781)
[PASS] testIncreaseAllowancewWithSuccess() (gas: 49878)
[PASS] testMintToZeroAddress() (gas: 25211)
[PASS] testMintWithNoAuthorization() (gas: 29629)
[PASS] testMintWithSuccess() (gas: 54137)
[PASS] testName() (gas: 9834)
[FAIL. Reason: Invalid data] testPermitExpiredDeadline() (gas: 13907)
Traces:
  [13907] SCTERC20Test::testPermitExpiredDeadline()
    ├─ [0] VM::2d0335ab(000000000000000000000000e1ab8145f7e55dc933d51a18c793f901a3a0b276)
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c496e76616c696420646174610000000000000000000000000000000000000000
    └─ ← "Invalid data"

[FAIL. Reason: Invalid data] testPermitInvalidSignature() (gas: 13886)
Traces:
  [13886] SCTERC20Test::testPermitInvalidSignature()
    ├─ [0] VM::2d0335ab(000000000000000000000000e1ab8145f7e55dc933d51a18c793f901a3a0b276)
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c496e76616c696420646174610000000000000000000000000000000000000000
    └─ ← "Invalid data"

[FAIL. Reason: Invalid data] testPermitToZeroAddress() (gas: 13864)
Traces:
  [13864] SCTERC20Test::testPermitToZeroAddress()
    ├─ [0] VM::2d0335ab(000000000000000000000000e1ab8145f7e55dc933d51a18c793f901a3a0b276)
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c496e76616c696420646174610000000000000000000000000000000000000000
    └─ ← "Invalid data"

[FAIL. Reason: Invalid data] testPermitWithSuccess() (gas: 13907)
Traces:
  [13907] SCTERC20Test::testPermitWithSuccess()
    ├─ [0] VM::2d0335ab(000000000000000000000000e1ab8145f7e55dc933d51a18c793f901a3a0b276)
    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c496e76616c696420646174610000000000000000000000000000000000000000
    └─ ← "Invalid data"

[PASS] testSymbol() (gas: 9922)
[PASS] testTotalSupply() (gas: 7613)
[PASS] testTransferFromInsufficientAllowance() (gas: 54421)
[PASS] testTransferFromInsufficientBalance() (gas: 43020)
[PASS] testTransferFromWithSuccess() (gas: 63939)
[PASS] testTransferToZeroAddress() (gas: 18078)
[PASS] testTransferWhenAmmountExceedsBalance() (gas: 22568)
[PASS] testTransferWhenBalanceIsZero() (gas: 22525)
[PASS] testTransferWithSuccess() (gas: 47342)
Test result: FAILED. 28 passed; 4 failed; finished in 10.50ms

Failed tests:
[FAIL. Reason: Invalid data] testPermitExpiredDeadline() (gas: 13907)
[FAIL. Reason: Invalid data] testPermitInvalidSignature() (gas: 13886)
[FAIL. Reason: Invalid data] testPermitToZeroAddress() (gas: 13864)
[FAIL. Reason: Invalid data] testPermitWithSuccess() (gas: 13907)

Encountered a total of 4 failing tests, 48 tests succeeded
```

freitasgouvea commented:
@jeffprestes I tried to reproduce the error you mentioned but I couldn't. I even deleted the repository and cloned a new one following the steps and again the tests worked here. If you can share more details about the error I can see if I can find it.

```shell
flavio@flavio-ubuntu:~$ git clone git@github.com:solid-world/solid-world-dao-contracts.git
Cloning into 'solid-world-dao-contracts'...
remote: Enumerating objects: 292, done.
remote: Counting objects: 100% (146/146), done.
remote: Compressing objects: 100% (72/72), done.
remote: Total 292 (delta 93), reused 100 (delta 70), pack-reused 146
Receiving objects: 100% (292/292), 1.48 MiB | 1.25 MiB/s, done.
Resolving deltas: 100% (155/155), done.
flavio@flavio-ubuntu:~$ cd solid-world-dao-contracts/
flavio@flavio-ubuntu:~/solid-world-dao-contracts$ git checkout feat/foundry-unit-tests
Branch 'feat/foundry-unit-tests' set up to track remote branch 'feat/foundry-unit-tests' from 'origin'.
Switched to a new branch 'feat/foundry-unit-tests'
flavio@flavio-ubuntu:~/solid-world-dao-contracts$ npm i

added 321 packages, and audited 322 packages in 3s

66 packages are looking for funding
  run `npm fund` for details

1 high severity vulnerability

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
flavio@flavio-ubuntu:~/solid-world-dao-contracts$ forge update lib/forge-std
Submodule 'lib/forge-std' (https://github.com/foundry-rs/forge-std) registered for path 'lib/forge-std'
Cloning into '/home/flavio/solid-world-dao-contracts/lib/forge-std'...
Submodule path 'lib/forge-std': checked out 'bdd321fc4ac81aff81924effb559427fcccca769'
Submodule 'lib/ds-test' (https://github.com/dapphub/ds-test) registered for path 'lib/forge-std/lib/ds-test'
Cloning into '/home/flavio/solid-world-dao-contracts/lib/forge-std/lib/ds-test'...
Submodule path 'lib/forge-std/lib/ds-test': checked out '2c7dbcc8586b33f358e3307a443e524490c17666'
flavio@flavio-ubuntu:~/solid-world-dao-contracts$ npm run test

> solid-world-dao-contracts@1.0.0 test
> forge test -vvv

[⠒] Compiling...
[⠘] Compiling 37 files with 0.8.13
[⠒] Solc finished in 4.59s
Compiler run successful (with warnings)

Running 20 tests for test/SolidDaoManagement.t.sol:SolidDaoManagementTest
[PASS] testGovernorAddress() (gas: 9909)
[PASS] testGuardianAddress() (gas: 9841)
[PASS] testPolicyAddress() (gas: 9799)
[PASS] testPullGovernor() (gas: 56316)
[PASS] testPullGuardian() (gas: 60349)
[PASS] testPullPolicy() (gas: 60296)
[PASS] testPullVault() (gas: 60468)
[PASS] testPushGovernor() (gas: 49631)
[PASS] testPushGuardian() (gas: 51667)
[PASS] testPushPolicy() (gas: 51675)
[PASS] testPushVault() (gas: 51696)
[PASS] testUnauthorizedPullGovernor() (gas: 47838)
[PASS] testUnauthorizedPullGuardian() (gas: 51762)
[PASS] testUnauthorizedPullPolicy() (gas: 51795)
[PASS] testUnauthorizedPullVault() (gas: 51818)
[PASS] testUnautorizedPushGovernor() (gas: 26150)
[PASS] testUnautorizedPushGuardian() (gas: 26194)
[PASS] testUnautorizedPushPolicy() (gas: 26193)
[PASS] testUnautorizedPushVault() (gas: 26235)
[PASS] testVaultAddress() (gas: 9951)
Test result: ok. 20 passed; 0 failed; finished in 10.24ms

Running 32 tests for test/SCTERC20.t.sol:SCTERC20Test
[PASS] testApproveFromZeroAddress() (gas: 15356)
[PASS] testApproveToZeroAddress() (gas: 15390)
[PASS] testApproveWithSuccess() (gas: 44516)
[PASS] testBalanceOf() (gas: 25531)
[PASS] testBurnFromInsufficientAllowance() (gas: 51874)
[PASS] testBurnFromInsufficientBalance() (gas: 40931)
[PASS] testBurnFromWithSuccess() (gas: 46932)
[PASS] testBurnWhenAmmountExceedsBalance() (gas: 16964)
[PASS] testBurnWithSuccess() (gas: 22458)
[PASS] testBurnWithZeroBalance() (gas: 16965)
[PASS] testBurnZeroAddress() (gas: 12538)
[PASS] testDecimals() (gas: 5502)
[PASS] testDecreaseAllowancewBellowZero() (gas: 45989)
[PASS] testDecreaseAllowancewWithSuccess() (gas: 49781)
[PASS] testIncreaseAllowancewWithSuccess() (gas: 49878)
[PASS] testMintToZeroAddress() (gas: 25211)
[PASS] testMintWithNoAuthorization() (gas: 29629)
[PASS] testMintWithSuccess() (gas: 54137)
[PASS] testName() (gas: 9834)
[PASS] testPermitExpiredDeadline() (gas: 29057)
[PASS] testPermitInvalidSignature() (gas: 55694)
[PASS] testPermitToZeroAddress() (gas: 53531)
[PASS] testPermitWithSuccess() (gas: 79591)
[PASS] testSymbol() (gas: 9922)
[PASS] testTotalSupply() (gas: 7613)
[PASS] testTransferFromInsufficientAllowance() (gas: 54421)
[PASS] testTransferFromInsufficientBalance() (gas: 43020)
[PASS] testTransferFromWithSuccess() (gas: 63939)
[PASS] testTransferToZeroAddress() (gas: 18078)
[PASS] testTransferWhenAmmountExceedsBalance() (gas: 22568)
[PASS] testTransferWhenBalanceIsZero() (gas: 22525)
[PASS] testTransferWithSuccess() (gas: 47342)
Test result: ok. 32 passed; 0 failed; finished in 10.24ms
```

jeffprestes commented:
Yes, I will try to get more info.
Meanwhile, could you please check the first comment in left in the review?

freitasgouvea commented:
> Yes, I will try to get more info. Meanwhile, could you please check the first comment in left in the review?

Great, I also added the instructions to only compile the smart contracts. 

And also It would be great to understand this error, it could even be a Foundry bug in certain contexts or environments.

Any other changes needed feel free to share.
